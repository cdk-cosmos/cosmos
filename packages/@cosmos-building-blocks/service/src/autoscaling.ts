import { Construct, Lazy, Fn } from '@aws-cdk/core';
import { InstanceType, InstanceClass, InstanceSize, CfnLaunchTemplate } from '@aws-cdk/aws-ec2';
import { CfnInstanceProfile } from '@aws-cdk/aws-iam';
import {
  CfnAutoScalingGroup,
  AutoScalingGroup,
  AutoScalingGroupProps,
  Monitoring,
  BlockDevice,
  EbsDeviceVolumeType,
  BlockDeviceVolume,
} from '@aws-cdk/aws-autoscaling';

/**
 * Properties for AutoScalingGroupLT
 */
export interface AutoScalingGroupLTProps extends AutoScalingGroupProps {
  /**
   * Set this to `true` if you want to use Instance Template instead of Launch Config.
   * You can fine tune your default LaunchTemplate with `instanceType`, `launchTemplateOverrides` and `instancesDistribution`
   * @default - false
   */
  readonly useInstanceTemplate?: boolean;
  /**
   * Type of instance to launch. if used with `useInstanceTemplate = true`, it will be applied to LaunchTemplate.
   * Make sure to check and tweak `launchTemplateOverrides` and `instancesDistribution` as the default values set for those
   * properties will overwrite this InstanceType
   */
  readonly instanceType: InstanceType;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Array of LaunchTemplateOverridesProperty which is a subproperty of LaunchTemplate that describes an override for a launch template.
   * Currently, the only supported override is instance type
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-launchtemplateoverrides.html
   * @default
   * [{ instanceType: "t3.medium" }, { instanceType: "t3.small" }]
   *
   * This will use any Spot instances available, "t3.medium" or "t3.small"
   * based on `spotAllocationStrategy` set in `instancesDistribution`
   */
  readonly launchTemplateOverrides?: Array<CfnAutoScalingGroup.LaunchTemplateOverridesProperty>;
  /**
   * Can only be used along with `useInstanceTemplate = true`.
   * Specifies the distribution of On-Demand Instances and Spot Instances,
   * the maximum price to pay for Spot Instances, and how the Auto Scaling group allocates instance types
   * to fulfill On-Demand and Spot capacity.
   * https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-instancesdistribution.html
   * @default - Uses all Spot instances of types specified in property `launchTemplateOverrides`
   *  {
   *    onDemandAllocationStrategy: "prioritized",
   *    onDemandBaseCapacity: 0,
   *    onDemandPercentageAboveBaseCapacity: 0,
   *    spotAllocationStrategy: "capacity-optimized",
   *    spotMaxPrice: props.spotPrice,
   *  }
   */
  readonly instancesDistribution?: Partial<CfnAutoScalingGroup.InstancesDistributionProperty>;
}

/**
 * Derived from Base class AutoScalingGroup with additional functionalities of using Launch Template (LT).
 *
 * Base AutoScalingGroup class uses Launch Config and have no ability to use more advanced Launch Templates
 * and MixedInstancesPolicies which gives ability to use Spot fleet and combination of spot and on-demand
 *
 * This construct uses default spot fleet template settings with just one property `useInstanceTemplate = true`
 *
 * With `launchTemplateOverrides` and `instancesDistribution`, it gives additional fine tuning.
 */
export class AutoScalingGroupLT extends AutoScalingGroup {
  constructor(scope: Construct, id: string, props: AutoScalingGroupLTProps) {
    super(scope, id, props);

    // Check if useInstanceTemplate is set to true. If not, do nothing and rely on Base class to
    // to create AutoScalingGroup using LaunchConfig
    const useInstanceTemplate = props.useInstanceTemplate !== undefined ? props.useInstanceTemplate === true : false;
    if (useInstanceTemplate == true) {
      // Prepare props required for LaunchTemplate
      const instanceType =
        props.instanceType !== undefined
          ? props.instanceType.toString()
          : InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString();

      const imageConfig = props.machineImage.getImage(this);
      const userDataToken = Lazy.stringValue({
        produce: () => Fn.base64(this.userData.render()),
      });
      const securityGroupIds = this.connections.securityGroups.map(sg => sg.securityGroupId);
      const instanceMonitoringMode =
        props.instanceMonitoring !== undefined ? props.instanceMonitoring === Monitoring.DETAILED : undefined;
      const monitoring = {
        enabled: instanceMonitoringMode == undefined ? true : instanceMonitoringMode,
      };

      const iamProfile = this.node.tryFindChild('InstanceProfile') as CfnInstanceProfile;

      // Create LaunchTemplate

      const lTemplate = new CfnLaunchTemplate(this, 'LaunchTemplate', {
        launchTemplateData: {
          imageId: imageConfig.imageId,
          securityGroupIds,
          userData: userDataToken,
          iamInstanceProfile: { arn: iamProfile.attrArn },
          keyName: props.keyName,
          monitoring,
          instanceType: instanceType,
          blockDeviceMappings:
            props.blockDevices !== undefined ? synthesizeBlockDeviceMappings(this, props.blockDevices) : undefined,
        },
      });
      lTemplate.node.addDependency(this.role);

      // Prepare props for ASG with MixedInstancesPolicy

      const overrides =
        props.launchTemplateOverrides == undefined
          ? [
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM).toString(),
              },
              {
                instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.SMALL).toString(),
              },
            ]
          : props.launchTemplateOverrides;

      const instancesDistribution = {
        onDemandAllocationStrategy: 'prioritized',
        onDemandBaseCapacity: 0,
        onDemandPercentageAboveBaseCapacity: 0,
        spotAllocationStrategy: 'capacity-optimized',
        spotMaxPrice: props.spotPrice,
        ...props.instancesDistribution,
      };

      // Mixed instance Policy
      const mixedInstancesPolicy: CfnAutoScalingGroup.MixedInstancesPolicyProperty = {
        launchTemplate: {
          launchTemplateSpecification: {
            version: lTemplate.attrLatestVersionNumber,
            launchTemplateId: lTemplate.ref,
          },
          overrides,
        },
        instancesDistribution,
      };

      // Remove LaunchConfig from the resulting CloudFormation template
      if (!this.node.tryRemoveChild('LaunchConfig')) throw new Error('Launch config not found');
      const asg = this.node.tryFindChild('ASG') as CfnAutoScalingGroup;
      asg.addPropertyDeletionOverride('LaunchConfigurationName');
      // Inject mixedInstancesPolicy to use LaunchTemplate
      asg.mixedInstancesPolicy = mixedInstancesPolicy;
    }
  }
}
function synthesizeBlockDeviceMappings(
  construct: Construct,
  blockDevices: BlockDevice[]
): CfnLaunchTemplate.BlockDeviceMappingProperty[] {
  return blockDevices.map<CfnLaunchTemplate.BlockDeviceMappingProperty>(({ deviceName, volume, mappingEnabled }) => {
    const { virtualName, ebsDevice: ebs } = volume;

    if (volume === BlockDeviceVolume.noDevice() || mappingEnabled === false) {
      return {
        deviceName,
        noDevice: 'true',
      };
    }

    if (ebs) {
      const { iops, volumeType } = ebs;

      if (!iops) {
        if (volumeType === EbsDeviceVolumeType.IO1) {
          throw new Error('iops property is required with volumeType: EbsDeviceVolumeType.IO1');
        }
      } else if (volumeType !== EbsDeviceVolumeType.IO1) {
        construct.node.addWarning('iops will be ignored without volumeType: EbsDeviceVolumeType.IO1');
      }
    }

    return {
      deviceName,
      ebs,
      virtualName,
    };
  });
}
