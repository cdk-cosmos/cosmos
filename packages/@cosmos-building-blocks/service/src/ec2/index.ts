export {
  CloudWatchAgent,
  CloudWatchAgentProps,
  CloudWatchMetricBuilder,
  CloudWatchMetric,
} from './ec2-cloud-watch-agent';
