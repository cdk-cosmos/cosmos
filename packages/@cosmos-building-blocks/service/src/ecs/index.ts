export { EcsService, EcsServiceProps } from './ecs-service';
