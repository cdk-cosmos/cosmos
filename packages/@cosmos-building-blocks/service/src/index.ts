export { EcsService, EcsServiceProps } from './ecs';
export { SecureBucket } from './s3';
export { Redis, IRedis } from './redis';
export { AutoScalingGroupV1Props, AutoScalingGroupV1 } from './autoscaling';
