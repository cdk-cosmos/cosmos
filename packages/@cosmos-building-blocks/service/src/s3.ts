import { Construct } from '@aws-cdk/core';
import { Bucket, BucketEncryption, BucketProps, BucketAccessControl } from '@aws-cdk/aws-s3';
import { PolicyStatement, Effect, AnyPrincipal } from '@aws-cdk/aws-iam';

export class SecureBucket extends Bucket {
  constructor(scope: Construct, id: string, props?: BucketProps) {
    super(scope, id, {
      encryption: BucketEncryption.S3_MANAGED,
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      serverAccessLogsPrefix: 'BucketAccessLogs/',
      ...props,
    });
    const policyStatement = new PolicyStatement({
      actions: ['s3:*'],
      principals: [new AnyPrincipal()],
      resources: [`${this.bucketArn}/*`],
      effect: Effect.DENY,
      sid: 'ForceSSLOnly',
    });
    policyStatement.addCondition('Bool', { 'aws:SecureTransport': false });
    super.addToResourcePolicy(policyStatement);
  }
}
