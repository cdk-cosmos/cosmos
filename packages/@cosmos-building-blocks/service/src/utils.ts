export const getRoutingPriority = (props: { hostHeader?: string; pathPattern?: string }): number => {
  const path = getStringValue(props.pathPattern);
  const host = getStringValue(props.hostHeader);
  const count = 2;
  const total = path + host;
  const inMin = 0 * count;
  const inMax = 128 * charMapping.length * count;
  const outMin = 1000; // Offset
  const outMax = Math.min(inMax + outMin, 50000); // only use and many as needed ?

  // Since lower numbers are more important in Listener Rules for ALB, then flip out min and max
  return mapNumber(total, inMin, inMax, outMax, outMin);
};

const mapNumber = (value: number, inMin: number, inMax: number, outMin: number, outMax: number): number => {
  return Math.floor(((value - inMin) / (inMax - inMin)) * (outMax - outMin) + outMin);
};

const getStringValue = (str?: string): number => {
  if (str === undefined) return 0;
  let total = 0;
  for (let i = 0; i <= str.length; i++) {
    total += charMapping.indexOf(str.charAt(i)) + 1;
  }
  return total;
};

const charMapping: string[] = [
  '*',
  '?',
  '&',
  '_',
  '@',
  '+',
  "'",
  '"',
  '~',
  '$',
  '.',
  ':',
  '-',
  'a',
  'b',
  'c',
  'd',
  'e',
  'f',
  'g',
  'h',
  'i',
  'j',
  'k',
  'l',
  'm',
  'n',
  'o',
  'p',
  'q',
  'r',
  's',
  't',
  'u',
  'v',
  'w',
  'x',
  'y',
  'z',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '0',
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '/',
];
