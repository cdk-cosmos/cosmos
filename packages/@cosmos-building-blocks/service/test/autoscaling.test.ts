import { SynthUtils, expect as cdkExpect, haveResourceLike, ABSENT, objectLike, arrayWith } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { AutoScalingGroupLT } from '../src/index';
import { Vpc, SubnetType, InstanceClass, InstanceType, InstanceSize, MachineImage } from '@aws-cdk/aws-ec2';
import { BlockDeviceVolume } from '@aws-cdk/aws-autoscaling';
import '@aws-cdk/assert/jest';

test('Should match snapshot', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'aws-autoscaling');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
    ],
  });
  // ASG using Launch Template
  new AutoScalingGroupLT(stack, 'SpotAsgLaunchTemplate', {
    useInstanceTemplate: true,
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
    spotPrice: '0.03',
  });

  // ASG using Launch Config
  new AutoScalingGroupLT(stack, 'DefaultAsgLaunchConfig', {
    useInstanceTemplate: false,
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
    // spotPrice: '0.03',
  });
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('When useInstanceTemplate flag set to true, should create ASG with default Launch Template settings', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'aws-autoscaling');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
    ],
  });
  const spotasg = new AutoScalingGroupLT(stack, 'DefaultSpotAsgLaunchTemplate', {
    useInstanceTemplate: true,
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
    spotPrice: '0.03',
  });
  spotasg.scaleOnCpuUtilization('CPUScale', { targetUtilizationPercent: 70 });

  const MixedInstancesPolicy = {
    InstancesDistribution: {
      OnDemandAllocationStrategy: 'prioritized',
      OnDemandBaseCapacity: 0,
      OnDemandPercentageAboveBaseCapacity: 0,
      SpotAllocationStrategy: 'capacity-optimized',
      SpotMaxPrice: '0.03',
    },
    LaunchTemplate: {
      LaunchTemplateSpecification: {
        LaunchTemplateId: {
          Ref: 'DefaultSpotAsgLaunchTemplate6D11B85A',
        },
        Version: {
          'Fn::GetAtt': ['DefaultSpotAsgLaunchTemplate6D11B85A', 'LatestVersionNumber'],
        },
      },
      Overrides: [
        {
          InstanceType: 't3.medium',
        },
        {
          InstanceType: 't3.small',
        },
      ],
    },
  };

  // THEN
  // ASG should have Launch Template and scaling policy
  cdkExpect(stack).to(haveResourceLike('AWS::EC2::LaunchTemplate'));
  cdkExpect(stack).to(haveResourceLike('AWS::AutoScaling::ScalingPolicy'));
  cdkExpect(stack).notTo(haveResourceLike('AWS::AutoScaling::LaunchConfiguration'));
  // ASG should have MixedInstancesPolicy
  cdkExpect(stack).to(
    haveResourceLike('AWS::AutoScaling::AutoScalingGroup', { MixedInstancesPolicy: MixedInstancesPolicy })
  );
});

test('Switching off or not passing LauchTemplate flag should create standard ASG from base AutoScalingGroup', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'autoscaling-default');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
    ],
  });
  const stack1 = new cdk.Stack(app, 'autoscaling-default1');
  const vpc1 = new Vpc(stack1, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
    ],
  });
  // ASG LT with only required props
  const spotasg = new AutoScalingGroupLT(stack, 'StandardASG', {
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
  });
  spotasg.scaleOnCpuUtilization('CPUScale', { targetUtilizationPercent: 70 });

  new AutoScalingGroupLT(stack1, 'StandardASGWithProp', {
    useInstanceTemplate: false, // Use standard @aws-cdk Base Class AutoScalingGroup
    machineImage: MachineImage.latestAmazonLinux(),
    vpc: vpc1,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
  });
  // THEN
  // ASG should have Launch Config and scaling policy
  cdkExpect(stack).to(haveResourceLike('AWS::AutoScaling::LaunchConfiguration'));
  cdkExpect(stack).to(haveResourceLike('AWS::AutoScaling::ScalingPolicy'));
  // ASG should have Launch Config and should not have LaunchTemplate
  cdkExpect(stack1).to(haveResourceLike('AWS::AutoScaling::LaunchConfiguration'));
  cdkExpect(stack1).notTo(haveResourceLike('AWS::EC2::LaunchTemplate'));
});

test('Should be able to overwrite LauchTemplate related properties', () => {
  //WHEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'aws-autoscaling');
  const vpc = new Vpc(stack, 'vpc', {
    maxAzs: 3,
    subnetConfiguration: [
      {
        name: 'App',
        subnetType: SubnetType.ISOLATED,
        cidrMask: 26,
      },
    ],
  });

  // Overwrite launchTemplateOverrides and instancesDistribution (onDemandBaseCapacity, spotAllocationStrategy)
  // Also overwrite blockDevices to check if new volume mapping will be created or not
  const spotasg = new AutoScalingGroupLT(stack, 'OverwriteLTASG', {
    useInstanceTemplate: true,
    launchTemplateOverrides: [{ instanceType: 'm2.small' }, { instanceType: 'm2.large' }],
    instancesDistribution: {
      onDemandBaseCapacity: 1,
      spotAllocationStrategy: 'lowest-price',
    },
    machineImage: MachineImage.latestAmazonLinux(),
    vpc,
    vpcSubnets: { subnetGroupName: 'App' },
    minCapacity: 0,
    maxCapacity: 5,
    desiredCapacity: 1,
    instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM),
    blockDevices: [{ volume: BlockDeviceVolume.ebs(20), deviceName: 'xvdb' }],
    spotPrice: '0.03',
  });
  spotasg.scaleOnCpuUtilization('CPUScale', { targetUtilizationPercent: 70 });

  // THEN
  // ASG should have Launch Template and scaling policy
  cdkExpect(stack).to(haveResourceLike('AWS::EC2::LaunchTemplate'));
  cdkExpect(stack).to(haveResourceLike('AWS::AutoScaling::ScalingPolicy'));
  cdkExpect(stack).notTo(haveResourceLike('AWS::AutoScaling::LaunchConfiguration'));
  // Launch Template should have new BlockDeviceMappings
  cdkExpect(stack).to(
    haveResourceLike('AWS::EC2::LaunchTemplate', {
      LaunchTemplateData: objectLike({
        BlockDeviceMappings: [
          {
            DeviceName: 'xvdb',
            Ebs: {
              VolumeSize: 20,
            },
          },
        ],
      }),
    })
  );
  // ASG should have overwritten properties
  cdkExpect(stack).to(
    haveResourceLike('AWS::AutoScaling::AutoScalingGroup', {
      LaunchConfigurationName: ABSENT,
      MixedInstancesPolicy: objectLike({
        InstancesDistribution: objectLike({ OnDemandBaseCapacity: 1, SpotAllocationStrategy: 'lowest-price' }),
        LaunchTemplate: objectLike({
          Overrides: arrayWith(
            {
              InstanceType: 'm2.small',
            },
            {
              InstanceType: 'm2.large',
            }
          ),
        }),
      }),
    })
  );
});
