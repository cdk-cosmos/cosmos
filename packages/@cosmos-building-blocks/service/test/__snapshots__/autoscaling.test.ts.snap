// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "DefaultAsgLaunchConfig8140DA5E": Object {
      "DependsOn": Array [
        "DefaultAsgLaunchConfigInstanceRoleF77373C9",
      ],
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "DefaultAsgLaunchConfigInstanceProfile1F634371",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.medium",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "DefaultAsgLaunchConfigInstanceSecurityGroup339317F1",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash",
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "DefaultAsgLaunchConfigASG06101099": Object {
      "Properties": Object {
        "DesiredCapacity": "1",
        "LaunchConfigurationName": Object {
          "Ref": "DefaultAsgLaunchConfig8140DA5E",
        },
        "MaxSize": "5",
        "MinSize": "0",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "aws-autoscaling/DefaultAsgLaunchConfig",
          },
        ],
        "VPCZoneIdentifier": Array [
          Object {
            "Ref": "vpcAppSubnet1Subnet0009B9D9",
          },
          Object {
            "Ref": "vpcAppSubnet2Subnet01033D46",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "DefaultAsgLaunchConfigInstanceProfile1F634371": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "DefaultAsgLaunchConfigInstanceRoleF77373C9",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "DefaultAsgLaunchConfigInstanceRoleF77373C9": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/DefaultAsgLaunchConfig",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DefaultAsgLaunchConfigInstanceSecurityGroup339317F1": Object {
      "Properties": Object {
        "GroupDescription": "aws-autoscaling/DefaultAsgLaunchConfig/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/DefaultAsgLaunchConfig",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "SpotAsgLaunchTemplate2147FC9C": Object {
      "DependsOn": Array [
        "SpotAsgLaunchTemplateInstanceRole62ACCAC2",
      ],
      "Properties": Object {
        "LaunchTemplateData": Object {
          "IamInstanceProfile": Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "SpotAsgLaunchTemplateInstanceProfileF66B715A",
                "Arn",
              ],
            },
          },
          "ImageId": Object {
            "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
          },
          "InstanceType": "t3.medium",
          "Monitoring": Object {
            "Enabled": true,
          },
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "SpotAsgLaunchTemplateInstanceSecurityGroupC1F6C06B",
                "GroupId",
              ],
            },
          ],
          "UserData": Object {
            "Fn::Base64": "#!/bin/bash",
          },
        },
      },
      "Type": "AWS::EC2::LaunchTemplate",
    },
    "SpotAsgLaunchTemplateASGB05FC1E5": Object {
      "Properties": Object {
        "DesiredCapacity": "1",
        "MaxSize": "5",
        "MinSize": "0",
        "MixedInstancesPolicy": Object {
          "InstancesDistribution": Object {
            "OnDemandAllocationStrategy": "prioritized",
            "OnDemandBaseCapacity": 0,
            "OnDemandPercentageAboveBaseCapacity": 0,
            "SpotAllocationStrategy": "capacity-optimized",
            "SpotMaxPrice": "0.03",
          },
          "LaunchTemplate": Object {
            "LaunchTemplateSpecification": Object {
              "LaunchTemplateId": Object {
                "Ref": "SpotAsgLaunchTemplate2147FC9C",
              },
              "Version": Object {
                "Fn::GetAtt": Array [
                  "SpotAsgLaunchTemplate2147FC9C",
                  "LatestVersionNumber",
                ],
              },
            },
            "Overrides": Array [
              Object {
                "InstanceType": "t3.medium",
              },
              Object {
                "InstanceType": "t3a.medium",
              },
            ],
          },
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "aws-autoscaling/SpotAsgLaunchTemplate",
          },
        ],
        "VPCZoneIdentifier": Array [
          Object {
            "Ref": "vpcAppSubnet1Subnet0009B9D9",
          },
          Object {
            "Ref": "vpcAppSubnet2Subnet01033D46",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "SpotAsgLaunchTemplateInstanceProfileF66B715A": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Ref": "SpotAsgLaunchTemplateInstanceRole62ACCAC2",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "SpotAsgLaunchTemplateInstanceRole62ACCAC2": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "ec2.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/SpotAsgLaunchTemplate",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "SpotAsgLaunchTemplateInstanceSecurityGroupC1F6C06B": Object {
      "Properties": Object {
        "GroupDescription": "aws-autoscaling/SpotAsgLaunchTemplate/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/SpotAsgLaunchTemplate",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "vpcA2121C38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcAppSubnet1RouteTable883411EE": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/vpc/AppSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcAppSubnet1RouteTableAssociation7C392A32": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcAppSubnet1RouteTable883411EE",
        },
        "SubnetId": Object {
          "Ref": "vpcAppSubnet1Subnet0009B9D9",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcAppSubnet1Subnet0009B9D9": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "App",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/vpc/AppSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcAppSubnet2RouteTable9E7B4A01": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/vpc/AppSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcAppSubnet2RouteTableAssociationD024F74B": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "vpcAppSubnet2RouteTable9E7B4A01",
        },
        "SubnetId": Object {
          "Ref": "vpcAppSubnet2Subnet01033D46",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcAppSubnet2Subnet01033D46": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.64/26",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "App",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
          Object {
            "Key": "Name",
            "Value": "aws-autoscaling/vpc/AppSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
